src/nlp/arrow_dataset.py

```diff
@@ -1284,7 +1322,7 @@ def shuffle(
         ), "The provided generator must be an instance of numpy.random.Generator"
 
         # Check if we've already cached this computation (indexed by a hash)
-        if self._data_files:
+        if self._data_files and (seed is not None or generator is not None):
```

Comment:
I think it would be better to simply remove all this `if` clause completely.
Indeed we can ignore this caching part as it is already present inside `select` that is called right after.

It also fixes the issue you mentioned.
Indeed when seed is None and generator is None, then it's going to be `permutation` that is going to influence the cache file name (that's how `select` works). Therefore if you call shuffle twice then it's going to return different datasets, as `permutation` will be different

Reply:
Actually I can see a case where this if clause has its advantages. It can avoid call `generator. permutation` and speed up things for huge datasets.

Therefore what I would suggest instead is simply move 

```python
        if generator is None:
            generator = np.random.default_rng(seed)
```

right before this if clause. It should to the job

Reply:
Done. Also removed `seed` from the cache in this case, since generator will always be defined and seed may not be.